README.md
# Projeto
 
API básica para treinamento e predição do modelo ML

# Dependências
Rodar 
```
pip install -r requirements.txt
```
# Execução 
Rodar
```
python run.py
```

history.txt
[PREDICT]  {'pir': 0, 'ldr_1': 0, 'ldr_2': 1, 'mic': 59, 'temp': 24.18, 'hum': 67.53, 'timestamp': 10, 'weekday': 4}
[PREDICT]  [[0.0]]
2023-05-04 10:36:47 [PREDICT]  LAMP OFF
2023-05-04 10:42:05 [PREDICT]  CP: 1 IP: 0
[PREDICT]  {'pir': 1, 'ldr_1': 0, 'ldr_2': 1, 'mic': 20, 'temp': 24.4, 'hum': 69.95, 'timestamp': 11, 'weekday': 4}
[PREDICT]  [[0.045873671770095825]]
2023-05-04 11:12:09 [PREDICT]  LAMP OFF
2023-05-04 11:17:09 [PREDICT]  CP: 2 IP: 0
[PREDICT]  {'pir': 0, 'ldr_1': 0, 'ldr_2': 1, 'mic': 0, 'temp': 24.25, 'hum': 68.62, 'timestamp': 11, 'weekday': 4}
[PREDICT]  [[0.017753561958670616]]
2023-05-04 11:47:12 [PREDICT]  LAMP OFF
2023-05-04 11:52:12 [PREDICT]  CP: 3 IP: 0
[PREDICT]  {'pir': 0, 'ldr_1': 0, 'ldr_2': 1, 'mic': 0, 'temp': 24.25, 'hum': 67.21, 'timestamp': 12, 'weekday': 4}
[PREDICT]  [[0.007761821616441011]]
2023-05-04 12:22:15 [PREDICT]  LAMP OFF
2023-05-04 12:27:15 [PREDICT]  CP: 4 IP: 0
[PREDICT]  {'pir': 1, 'ldr_1': 0, 'ldr_2': 1, 'mic': 3, 'temp': 24.26, 'hum': 66.27, 'timestamp': 12, 'weekday': 4}
[PREDICT]  [[0.09280405193567276]]
2023-05-04 12:57:18 [PREDICT]  LAMP OFF
2023-05-04 13:02:18 [PREDICT]  CP: 5 IP: 0
[PREDICT]  {'pir': 1, 'ldr_1': 0, 'ldr_2': 1, 'mic': 0, 'temp': 24.32, 'hum': 66.14, 'timestamp': 13, 'weekday': 4}
[PREDICT]  [[0.010709145106375217]]
2023-05-04 13:32:21 [PREDICT]  LAMP OFF
2023-05-04 13:37:21 [PREDICT]  CP: 6 IP: 0
[PREDICT]  {'pir': 1, 'ldr_1': 0, 'ldr_2': 1, 'mic': 4, 'temp': 24.28, 'hum': 65.78, 'timestamp': 14, 'weekday': 4}
[PREDICT]  [[0.07271843403577805]]
2023-05-04 14:07:24 [PREDICT]  LAMP OFF
2023-05-04 14:12:24 [PREDICT]  CP: 7 IP: 0
[PREDICT]  {'pir': 0, 'ldr_1': 0, 'ldr_2': 1, 'mic': 192, 'temp': 24.33, 'hum': 65.73, 'timestamp': 14, 'weekday': 4}
[PREDICT]  [[0.0]]
2023-05-04 14:42:27 [PREDICT]  LAMP OFF
2023-05-04 14:47:27 [PREDICT]  CP: 8 IP: 0
[PREDICT]  {'pir': 1, 'ldr_1': 0, 'ldr_2': 1, 'mic': 5, 'temp': 24.38, 'hum': 65.35, 'timestamp': 15, 'weekday': 4}
[PREDICT]  [[0.10943833738565445]]
2023-05-04 15:17:30 [PREDICT]  LAMP OFF
2023-05-04 15:22:30 [PREDICT]  CP: 9 IP: 0
[PREDICT]  {'pir': 1, 'ldr_1': 0, 'ldr_2': 1, 'mic': 25, 'temp': 24.44, 'hum': 65.8, 'timestamp': 15, 'weekday': 4}
[PREDICT]  [[0.16412347555160522]]
2023-05-04 15:52:32 [PREDICT]  LAMP OFF
2023-05-04 15:57:33 [PREDICT]  CP: 10 IP: 0
[PREDICT]  {'pir': 1, 'ldr_1': 0, 'ldr_2': 1, 'mic': 0, 'temp': 24.4, 'hum': 65.97, 'timestamp': 16, 'weekday': 4}
[PREDICT]  [[0.0013061241479590535]]
2023-05-04 16:27:35 [PREDICT]  LAMP OFF
2023-05-04 16:32:35 [PREDICT]  CP: 11 IP: 0
[PREDICT]  {'pir': 1, 'ldr_1': 0, 'ldr_2': 1, 'mic': 0, 'temp': 24.4, 'hum': 66.27, 'timestamp': 17, 'weekday': 4}
[PREDICT]  [[0.0006467806524597108]]
2023-05-04 17:02:38 [PREDICT]  LAMP OFF
2023-05-04 17:07:38 [PREDICT]  CP: 12 IP: 0
[PREDICT]  {'pir': 1, 'ldr_1': 0, 'ldr_2': 1, 'mic': 0, 'temp': 24.41, 'hum': 66.28, 'timestamp': 17, 'weekday': 4}
[PREDICT]  [[0.0006469322834163904]]
2023-05-04 17:37:41 [PREDICT]  LAMP OFF
2023-05-04 17:42:41 [PREDICT]  CP: 13 IP: 0
[PREDICT]  {'pir': 1, 'ldr_1': 1, 'ldr_2': 1, 'mic': 35, 'temp': 24.43, 'hum': 66.56, 'timestamp': 18, 'weekday': 4}
[PREDICT]  [[0.8858500123023987]]
2023-05-04 18:12:44 [PREDICT]  LAMP ON
2023-05-04 18:17:44 [PREDICT]  CP: 14 IP: 0
[PREDICT]  {'pir': 1, 'ldr_1': 1, 'ldr_2': 1, 'mic': 19, 'temp': 24.44, 'hum': 67.0, 'timestamp': 18, 'weekday': 4}
[PREDICT]  [[0.015344909392297268]]
2023-05-04 18:47:47 [PREDICT]  LAMP OFF
2023-05-04 18:52:52 [PREDICT]  CP: 14 IP: 1
[PREDICT]  {'pir': 0, 'ldr_1': 1, 'ldr_2': 1, 'mic': 439, 'temp': 24.55, 'hum': 68.44, 'timestamp': 19, 'weekday': 4}
[PREDICT]  [[0.0]]
2023-05-04 19:22:55 [PREDICT]  LAMP OFF
2023-05-04 19:27:59 [PREDICT]  CP: 14 IP: 2
[PREDICT]  {'pir': 0, 'ldr_1': 1, 'ldr_2': 1, 'mic': 0, 'temp': 24.53, 'hum': 68.9, 'timestamp': 20, 'weekday': 4}
[PREDICT]  [[4.970142981619574e-07]]
2023-05-04 20:15:16 [PREDICT]  LAMP OFF
2023-05-04 20:20:21 [PREDICT]  CP: 14 IP: 3
[PREDICT]  {'pir': 1, 'ldr_1': 1, 'ldr_2': 1, 'mic': 77, 'temp': 24.5, 'hum': 69.07, 'timestamp': 20, 'weekday': 4}
[PREDICT]  [[0.0]]
2023-05-04 20:45:18 [PREDICT]  LAMP OFF
2023-05-04 20:50:23 [PREDICT]  CP: 14 IP: 4
[PREDICT]  {'pir': 1, 'ldr_1': 1, 'ldr_2': 1, 'mic': 0, 'temp': 24.5, 'hum': 69.17, 'timestamp': 21, 'weekday': 4}
[PREDICT]  [[5.704245040760725e-07]]
2023-05-04 21:15:21 [PREDICT]  LAMP OFF
2023-05-04 21:20:26 [PREDICT]  CP: 14 IP: 5
[PREDICT]  {'pir': 1, 'ldr_1': 1, 'ldr_2': 1, 'mic': 4, 'temp': 24.62, 'hum': 69.93, 'timestamp': 21, 'weekday': 4}
[PREDICT]  [[8.173997230187524e-06]]
2023-05-04 21:45:24 [PREDICT]  LAMP OFF
2023-05-04 21:50:25 [PREDICT]  CP: 15 IP: 5
[PREDICT]  {'pir': 1, 'ldr_1': 1, 'ldr_2': 1, 'mic': 3, 'temp': 24.65, 'hum': 70.29, 'timestamp': 22, 'weekday': 4}
[PREDICT]  [[1.6507384543729131e-06]]
2023-05-04 22:15:27 [PREDICT]  LAMP OFF
2023-05-04 22:20:32 [PREDICT]  CP: 15 IP: 6
[PREDICT]  {'pir': 1, 'ldr_1': 1, 'ldr_2': 1, 'mic': 7, 'temp': 24.59, 'hum': 70.09, 'timestamp': 22, 'weekday': 4}
[PREDICT]  [[0.00022064401127863675]]
2023-05-04 22:45:30 [PREDICT]  LAMP OFF
2023-05-04 22:50:35 [PREDICT]  CP: 15 IP: 7
[PREDICT]  {'pir': 0, 'ldr_1': 1, 'ldr_2': 1, 'mic': 4, 'temp': 24.53, 'hum': 69.97, 'timestamp': 23, 'weekday': 4}
[PREDICT]  [[1.2391321888571838e-06]]
2023-05-04 23:15:32 [PREDICT]  LAMP OFF
2023-05-04 23:20:37 [PREDICT]  CP: 15 IP: 8
[PREDICT]  {'pir': 1, 'ldr_1': 1, 'ldr_2': 1, 'mic': 128, 'temp': 24.59, 'hum': 70.46, 'timestamp': 23, 'weekday': 4}
[PREDICT]  [[0.0]]
2023-05-04 23:45:35 [PREDICT]  LAMP OFF
2023-05-04 23:50:40 [PREDICT]  CP: 15 IP: 9
[PREDICT]  {'pir': 0, 'ldr_1': 1, 'ldr_2': 1, 'mic': 2, 'temp': 24.59, 'hum': 70.35, 'timestamp': 0, 'weekday': 5}
[PREDICT]  [[0.007327504456043243]]
2023-05-05 00:15:38 [PREDICT]  LAMP OFF
2023-05-05 00:20:43 [PREDICT]  CP: 15 IP: 10
[PREDICT]  {'pir': 0, 'ldr_1': 1, 'ldr_2': 1, 'mic': 4, 'temp': 24.56, 'hum': 70.03, 'timestamp': 0, 'weekday': 5}
[PREDICT]  [[0.0018454899545758963]]

scheduler.py
import schedule
import time
import os
import datetime
import requests
import pandas as pd
import threading

BANANA_API="http://192.168.0.43"
SENSORS_API="http://192.168.0.35"
LAMP_API="http://192.168.0.46"

CORRECT_PRED = 0
INCORRECT_PRED = 0

os.environ['TZ'] = 'Brazil/East'
time.tzset()

def check_prediction(sensor_state: dict, prediction: float):
  global CORRECT_PRED, INCORRECT_PRED
  try:
    res = requests.get(f"{LAMP_API}/cm?cmnd=Power", timeout=60).json()

    if res["POWER"] == "ON" and prediction >= 0.5 or \
      res["POWER"] == "OFF" and prediction < 0.5:
      CORRECT_PRED += 1
    else:
      # Save incorrect prediction
      INCORRECT_PRED += 1
      requests.post(f"{BANANA_API}:5000/peripherals", json=sensor_state, timeout=100)

    print(f'{time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())} [PREDICT] ', f"CP: {CORRECT_PRED} IP: {INCORRECT_PRED}", flush=True)
  except Exception as e:
    print(f'{time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())} [PREDICT] ', e, flush=True)


def predict():
  global CORRECT_PRED, INCORRECT_PRED
  try:
    sensors_state = requests.get(f"{SENSORS_API}", timeout=10).json()
    sensors = sensors_state.copy()

    sensors["timestamp"] = int(time.strftime('%H', time.localtime()))
    sensors["weekday"] = int(time.strftime('%w', time.localtime()))

    pred = requests.post(f"{BANANA_API}:3000/predict", json=[sensors]).json()

    print("[PREDICT] ", sensors, flush=True)
    print("[PREDICT] ", pred, flush=True)

    if pred[0][0] >= 0.5:
      requests.get(f"{LAMP_API}/cm?cmnd=Power%20ON", timeout=60)
      print(f'{time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())} [PREDICT] ', "LAMP ON", flush=True)
    else:
      requests.get(f"{LAMP_API}/cm?cmnd=Power%20OFF", timeout=60)
      print(f'{time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())} [PREDICT] ', "LAMP OFF", flush=True)
  
    # Check prediction
    threading.Timer(5*60, lambda: check_prediction(sensors_state, pred[0][0])).start()

  except Exception as e:
    print(f'{time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())} [PREDICT] ', e, flush=True)

def retrain():
  global CORRECT_PRED, INCORRECT_PRED
  try:
    start = int((datetime.datetime.today() - datetime.timedelta(days=3)).timestamp())
    end = int(datetime.datetime.today().timestamp())

    data = pd.DataFrame(requests.get(f"{BANANA_API}:5000/peripherals?start={start}&end={end}", timeout=100).json())
    
    if data.empty:
      print("[TRAIN] NO DATA", flush=True)
      return

    # Extract target
    target = data.pop("led_on")

    # Drop led color cols
    data.drop(columns=["led_color", "led_dimmer"], inplace=True)

    # Drop rows with negative values
    data = data[data["temp"]>=0]
    data = data[data["hum"]>=0]
    data = data[data["mic"]>=0]

    # Turn timestamp integer to category type based on hour
    data["weekday"] = data["timestamp"].apply(lambda x: int(time.strftime('%w', time.localtime(x))))
    data["timestamp"] = data["timestamp"].apply(lambda x: int(time.strftime('%H', time.localtime(x))))

    res = requests.post(f"{BANANA_API}:3000/train", json={"x": data.to_dict('records'), "y": target.tolist(), "cp": CORRECT_PRED, "ip": INCORRECT_PRED})

    # Reset counters
    CORRECT_PRED = 0
    INCORRECT_PRED = 0

    print(f'{time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())} [TRAIN] DONE - RES {res.status_code}', flush=True)
  except Exception as e:
    print(f'{time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())} [TRAIN] ', e, flush=True)

schedule.every(3).days.at("00:15").do(retrain)
schedule.every(30).minutes.do(predict)

while True:
  # Checks whether a scheduled task
  # is pending to run or not
  schedule.run_pending()
  time.sleep(10)